okhttp/src/main/kotlin/okhttp3/internal/connection/RealConnection.kt

```diff
@@ -146,25 +146,21 @@ class RealConnection(
    * Returns true if this is an HTTP/2 connection. Such connections can be used in multiple HTTP
    * requests simultaneously.
    */
-  val isMultiplexed: Boolean
+  internal val isMultiplexed: Boolean
     get() = http2Connection != null
 
   /** Prevent further exchanges from being created on this connection. */
-  fun noNewExchanges() {
-    assertThreadDoesntHoldLock()
```

Comment:
This is the one line that concerns me.  We don't use a static analysis check like GuardedBy, and these particular checks always gave me more confidence that the somewhat complicated and fluid concurrency dance we attempt doesn't have some potential race conditions or theoretical deadlocks.

Reply:
The assertion is most useful in a few cases:
 - checking a different lock that it's acquiring (possible deadlock)
 - checking no lock before doing I/O

I wanna keep the check in these situations. The check is doing at runtime what I'd prefer a framework or type checker to do for us statically.

In these cases it's a pretty benign check: are we using reentrant capabilities of the lock. I find reentrant synchronization to be quite sloppy, and an invitation for bugs. But there's no particular danger in these changed cases; we’re not doing any deadlock-capable locking or I/O.

Your comment does suggest something we should fix though! We have a check for unexpected locks held in our EventListener. I should update that to check for locks on connections also!